### ISSUE TITLE
Two race conditions (findOrCreate): cannot create concurrent transactions; and errors lead to wrong connection uuid

### ISSUE THREAD

**Issue by Domiii** [sequelize/sequelize#13554]
## Issue Creation Checklist

[x] I have read the [contribution guidelines](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)

## Bug Description

This issue is a continuation of the story from #1831. There are two more bugs:

1. Cannot execute multiple transactions concurrently.
   * Since #1831 was fixed, `findOrCreate` is wrapped in a transaction.
   * But most dialects do not support more than one transaction active at a time.
   * Thus the following code featuring concurrent `findOrCreate` operations bugs out: `SQLITE` reports an error when trying to create the second transaction while the first is still ongoing.
   * This needs to be dealt with more gracefully, as it causes further bugs down the line.
2. The above error causes the 2nd transaction's `cleanup` to reset the 1st transaction on global connection object.
   * When the second `findOrCreate` failed, it calls `rollback` → `transaction.cleanup` which in turn unsets `connection.uuid`.
   * The query log now shows that the `create` part of the first `findOrCreate` goes against the global connection ("default"), and not against its own transaction anymore.
     * The transaction is now gone and even the first operation has no transaction anymore.
     * Meaning: it succeeds but without atomicity guarantees!
   * In a final twist, since the transaction has already been cleaned up, the final COMMIT of the first operation also fails.
     * If you use `Promise.all` here, this second error message (coming from the first transaction) actually gets swallowed. You won't even see anything in the console, as is explained [here](https://dev.to/domiii/the-dangling-fallacy-with-promiseall-543n).
     * It does not actually cause obvious problems, since the operation has already succeeded, but still: it just shows how many small problems arise from that transaction getting pulled from right under the operation's feet:
     * Error message:
       ```
       Error: SQLITE_ERROR: cannot commit - no transaction is active
       at formatError (code\sequelize\lib\dialects\sqlite\query.js:403:16)
       at _handleQueryResponse (code\sequelize\lib\dialects\sqlite\query.js:72:13)     
       at Statement.afterExecute (code\sequelize\lib\dialects\sqlite\query.js:238:21)  
       ```

### SSCCE

**Here is the link to the SSCCE for this issue:** https://github.com/papb/sequelize-sscce/pull/189

```js
/* eslint-disable no-console */
'use strict';

const path = require('path');
const { expect } = require('chai');
const { createSequelizeInstance } = require('./dev/sscce-helpers');
const { Model, DataTypes } = require('.');

class User extends Model { }

async function main() {
  try {
    const sequelize = createSequelizeInstance({
      benchmark: true,
      retry: {
        max: 0
      }
    });
    User.init({
      name: DataTypes.STRING,
      age: DataTypes.INTEGER
    }, { sequelize, modelName: 'user' });
    await sequelize.sync({ force: true });

    // this does not work (AV)
    const [b1, b2] = await Promise.all([
      User.findOrCreate({
        where: { name: "b" },
        defaults: { age: 1 }
      }),
      User.findOrCreate({
        where: { name: "b" },
        defaults: { age: 2 }
      })
    ]);
    console.log(`Result: ${[b1?.[0]?.dataValues?.age, b2?.[0]?.dataValues?.age] || 'undefined'} === 1, 1`);


    await sequelize.close();
  }
  catch (err) {
    console.error('####### FAIL\n\n', err);
    const result = await User.findAll({
      where: { name: "b" }
    });
    console.log('##### RESULT:', result);
  }
}

main();
```

### What do you expect to happen?

**Option 1**: Ideally: two transactions should be allowed to run in parallel, and without a problem.

**Option 2**: If the DBMS does not allow for that, parallel transactions should be serialized with a hidden queue, so it can at least succeed, even without proper concurrency.

**Option 3**: Less ideal: At least, the first transaction should work as expected and be left untouched by the second failing.

### What is actually happening?

```
Executed (default): CREATE TABLE IF NOT EXISTS `users` (`id` INTEGER PRIMARY KEY AUTOINCREMENT, `name` VARCHAR(255), `age` INTEGER, `createdAt` DATETIME NOT NULL, `updatedAt` DATETIME NOT NULL); Elapsed time: 13999ms
Executed (default): PRAGMA INDEX_LIST(`users`) Elapsed time: 8ms
Executed (540b8ffb-a000-49c1-86b7-dcbce42ab11b): BEGIN DEFERRED TRANSACTION; Elapsed time: 53ms
Executed (540b8ffb-a000-49c1-86b7-dcbce42ab11b): BEGIN DEFERRED TRANSACTION; Elapsed time: 412ms
Executed (540b8ffb-a000-49c1-86b7-dcbce42ab11b): ROLLBACK; Elapsed time: 17ms
####### FAIL

 code\sequelize\lib\dialects\sqlite\query.js:403
        return new sequelizeErrors.DatabaseError(err);
               ^

Error: SQLITE_ERROR: cannot start a transaction within a transaction
    at formatError (code\sequelize\lib\dialects\sqlite\query.js:403:16)
    at _handleQueryResponse (code\sequelize\lib\dialects\sqlite\query.js:72:13)     
    at Statement.afterExecute (code\sequelize\lib\dialects\sqlite\query.js:238:21)  
Executed (540b8ffb-a000-49c1-86b7-dcbce42ab11b): SELECT `id`, `name`, `age`, `createdAt`, `updatedAt` FROM `users` AS 
`user` WHERE `user`.`name` = 'b' LIMIT 1; Elapsed time: 428ms
Executed (default): INSERT INTO `users` (`id`,`name`,`age`,`createdAt`,`updatedAt`) VALUES (NULL,$1,$2,$3,$4); {"$1":"b","$2":1,"$3":"2021-10-11 21:03:31.259 +00:00","$4":"2021-10-11 21:03:31.259 +00:00"} Elapsed time: 25ms
Executed (default): SELECT `id`, `name`, `age`, `createdAt`, `updatedAt` FROM `users` AS `user` WHERE `user`.`name` = 
'b'; Elapsed time: 319ms
##### RESULT: [
  user { [(omitted for brevity)] } 
]
Executed (default): COMMIT; Elapsed time: 54ms
```

### Environment

- Sequelize version: v6.6.5 
- Node.js version: v16.* 

### How does this problem relate to dialects?

- [x] I think this problem happens regardless of the dialect. (The part where a failing transaction causes havoc on other concurrent queries happens regardless on dialect.)
- [ ] I think this problem happens only for the following dialect(s): 
- [x] I don't know, I was using sqlite3 (The part where only one transaction may be active at a time might be sqlite only.)

### Would you be willing to resolve this issue by submitting a Pull Request?

- [ ] Yes, I have the time and I know how to start.
- [ ] Yes, I have the time but I don't know how to start, I would need guidance.
- [x] No, I don't have the time, although I believe I could do it if I had the time... :(
- [ ] No, I don't have the time and I wouldn't even know how to start.

**Comment by Domiii** [comment]
Bump

**Comment by AntonPetrochenko** [comment]
Bumping for relevancy: still happens as described here

sqlite3-5.0.11
sequelize-6.21.4

**Comment by Domiii** [comment]
@AntonPetrochenko - My thoughts are:

1. We cannot currently run multiple `findOrCreate`s concurrently on the same connection, simply because SQLITE does not support it. Even with MYSQL, trying to start a second transaction will first commit any pending transaction.
2. If you want this to work, you probably either have to use an async queue (e.g. [the queue from async](https://caolan.github.io/async/v3/docs.html#queue) with `concurrency` set to `1`), or multiple separate connection objects.
